lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,94 @@
+class LRUHash
+  # This LRU (Least Recently Used) hash will allow
+  # the cleaning of resources as time goes.
+  # The  goal is to remove the least recently used resources
+  # everytime we set a new resource. A default window of
+  # 5 minutes will allow empty item to stay in the hash
+  # for a maximum of 5 minutes
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?
+  attr_reader :table, :minimum_time_in_lru
+  MINIMUM_TIME_IN_LRU = 300
+
+  class NoopMutex
+    def synchronize(*)
+      yield
+    end
+
+    def try_lock
+      true
+    end
+
+    def unlock
+      true
+    end
+  end
+
+  [:keys, :clear].each do |attribute|
+    define_method :"#{attribute}" do
+      @lock.synchronize { @table.public_send(attribute) }
+    end
+  end
+
+  def initialize
+    @minimum_time_in_lru = MINIMUM_TIME_IN_LRU
+    @table = {}
+    @lock =
+      if Semian.thread_safe?
+        Mutex.new
+      else
+        NoopMutex.new
+      end
+  end
+
+  def set(key, resource)
+    @lock.synchronize do
+      @table.delete(key)
+      @table[key] = resource
+    end
+    resource.updated_at = Time.now
+    clear_unused_resources
+  end
+
+  def get(key)
+    @lock.synchronize do
+      found = @table.delete(key)
+      if found
+        @table[key] = found
+      end
+    end
+    @table[key]
+  end
+
+  def delete(key)
+    @lock.synchronize do
+      @table.delete(key)
+    end
+  end
+
+  def []=(key, resource)
+    set(key, resource)
+  end
+
+  def [](key)
+    get(key)
+  end
+
+  private
+
+  def clear_unused_resources
+    # Clears resources that have not been used in the last 5 minutes.
+    Semian.notify(:lru_hash_cleaned, self, :cleaning, :lru_hash)
+    return unless @lock.try_lock
+    @table.each do |_, resource|
+      break if resource.updated_at + minimum_time_in_lru > Time.now
+      next if resource.in_use?
+
+      resource = @table.delete(resource.name)
+      if resource
+        resource.destroy
+      end
+    end
+    @lock.unlock
```

Comment:
I think this needs to be in an `ensure` block to avoid accidental lock poisoning.

Reply:
Unfortunately I think this is going to raise an exception.

> Raises ThreadError if mutex wasn't locked by the current thread.

I think an `if @lock.locked?` will do the trick.